### PR TITLE
Core: Support registerTable with REST session catalog

### PR DESCRIPTION
**Pull request by krvikash**
Core: Support `registerTable` with REST session catalog

### CODE REVIEW

**Comment by krvikash** [comment]
CI failure will be fixed with https://github.com/apache/iceberg/pull/6511

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java [hunk 1]
```diff
@@ -146,6 +149,21 @@ public void doCommit(TableMetadata base, TableMetadata metadata) {
     }
   }
 
+  private Optional<String> getExistingMetadataLocation(TableMetadata base, TableMetadata metadata) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java:152]: can you elaborate why these changes are needed in order to support registering a table in the REST catalog? I think stuff should work fine without these changes

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 2]
```diff
@@ -351,7 +354,35 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
...
+
+    return buildTable(context, ident, metadata.schema())
+        .withLocation(metadata.location())
+        .withProperties(tableProperties)
+        .withPartitionSpec(metadata.spec())
+        .withSortOrder(metadata.sortOrder())
+        .create();
+  }
+
+  private boolean isValidIdentifier(TableIdentifier tableIdentifier) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:383]: do we need this, given that it always returns true? I know we do the same in `BaseMetastoreCatalog` where catalog implementations can override this behavior but I don't think we need this here in particular. 
You might rather want to call `checkIdentifierIsValid(..)` in this case.

**Review by nastra — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 3]
```diff
@@ -351,7 +354,35 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
   @Override
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:354]: Just wondering why can't we remove this method and let it use the parent implementation.
Other catalogs do the same.

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:354]: oh.. This is `SessionCatalog`, not the `Catalog`. So, there is no parent implementation. 
I think this logic can be moved to `BaseSessionCatalog`. So, it can help if there are some more implementations in the future.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java (hunk 1 above)
Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java:152]: Yes, You are right. without this change also register table will work. But the `registerTable` will create a new metadata JSON file and register the table. That is why this handling ensures if `registerTable` call `doCommit` then don't create a new metadata file and rather use the existing one.

Currently we are handling the same logic in `HiveTableOperations` and `NessieTableOperations`. But other table operations are missing this part.

https://github.com/apache/iceberg/blob/6b8f7e0e31a81029b478e7757aba749f5ed27f42/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java#L249-L252

https://github.com/apache/iceberg/blob/ede085d0f7529f24acd0c81dd0a43f7bb969b763/nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java#L146-L149


 
Coming to `getExistingMetadataLocation`, I am setting `metadataFileLocation` in the meadata properties in `RESTSessionCatalog` whereas for `JdbcCatalog` it uses `BaseMetastoreCatalog` implementation where `metadata` with existing `metadataFileLocation` is passed through table operations.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 2 above)
Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:383]: Yes, I have copied `isValidIdentifier` from `BaseMetastoreCatalog` implementation. Thanks for the input. I will use `checkIdentifierIsValid` over here.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 3 above)
Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:354]: Thanks, @ajantha-bhat for the input. Yes, moving implementation logic makes sense to `BaseSessionCatalog`. I will update the code.

Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:354]: I see that all the method's implementation is in `RESTSessionCatalog.java`. IMO we should keep `registerTable` implementation in `RESTSessionCatalog.java` as well. WDYT?

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java (hunk 1 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java:152]: I would probably do these changes independently from this PR as the scope of this PR should be only around supporting table registration for the REST catalog

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 4]
```diff
@@ -351,7 +354,30 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
...
   public Table registerTable(
       SessionContext context, TableIdentifier ident, String metadataFileLocation) {
-    throw new UnsupportedOperationException("Register table is not supported");
+    checkIdentifierIsValid(ident);
+
+    Preconditions.checkArgument(
+        metadataFileLocation != null && !metadataFileLocation.isEmpty(),
+        "Cannot register an empty metadata file location as a table");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:361]: If the metadata location is null, this should not state that it was empty. Instead, use the more standard form for error messages: `"Invalid metadata file location: %s"`. Printing the location shows that it is null or empty.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 5]
```diff
@@ -351,7 +354,30 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
...
+            .putAll(metadata.properties())
+            .put(BaseMetastoreTableOperations.METADATA_LOCATION_PROP, metadataFileLocation)
+            .build();
+
+    return buildTable(context, ident, metadata.schema())
+        .withLocation(metadata.location())
+        .withProperties(tableProperties)
+        .withPartitionSpec(metadata.spec())
+        .withSortOrder(metadata.sortOrder())
+        .create();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:380]: This discards all table data and older schemas, specs, and sort orders. I think this needs to preserve the table metadata completely.

**Comment by krvikash** [comment]
> If we want to support register, then I think the protocol itself should support register

Hi @rdblue, Thanks for the review. If I understood correctly, you wanted to introduce a new `RESTRequest` for `registerTable` similar to what we have `CreateTableRequest`. I have introduced a new `RegisterTableRequest` and used it for `regsiterTable`. I have updated the test case as well where I am verifying verify metadata like specs, sortOrders, history, snapshots ...

Please let me know if this is not what you intended to convey.

**Comment by krvikash** [comment]
CI Failure will be fixed by https://github.com/apache/iceberg/pull/6589

```
TestJdbcCatalog > testRegisterTable() FAILED
    org.junit.ComparisonFailure: Name must match expected:<[test_jdbc_catalog.]newdb.table> but was:<[]newdb.table>
        at org.junit.Assert.assertEquals(Assert.java:117)
```

```
TestNessieCatalog > testRegisterTable() FAILED
    org.junit.ComparisonFailure: Name must match expected:<ne[ssie.ne]wdb.table> but was:<ne[]wdb.table>
        at org.junit.Assert.assertEquals(Assert.java:117)
```

**Comment by krvikash** [comment]
Rebased with master and resolved conflicts.

**Comment by krvikash** [comment]
Hi @ajantha-bhat | @nastra | @rdblue, when you get time, could you please review it again?

**Comment by nastra** [comment]
@krvikash looks like CI is failing on this one

**Comment by krvikash** [comment]
Thanks, @nastra for the reminder. I have fixed the CI failure.

**Comment by krvikash** [comment]
Hi @rdblue | @nastra | @ajantha-bhat , When you get time, could you please review this?

**Comment by krvikash** [comment]
Addressed comments

**Comment by krvikash** [comment]
rebased with master and resolved conflicts.

**Comment by krvikash** [comment]
Fixed CI failure.

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.requests;
+
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.rest.RESTRequest;
+
+public class RegisterTableRequest implements RESTRequest {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java:25]: @rdblue I'm not sure if we decided on which pattern we'd want to use with new request/response classes:
1) class with a public constructor and no explicit JSON parser + manually written builder
2) Immutable class (similar to `ReportMetricsRequest`) with a separate JSON parser

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 7]
```diff
@@ -2487,6 +2488,88 @@ public void tableCreationWithoutNamespace() {
...
+
+  @Test
+  public void testRegisterExistingTable() {
+    C catalog = catalog();
+
+    TableIdentifier identifier = TableIdentifier.of("a", "t1");
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(identifier.namespace());
+    }
+    catalog.createTable(identifier, SCHEMA);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2563]: nit: space missing after }

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java (hunk 6 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java:25]: @krvikash since we don't want to rely on reflection (anymore) and want to be explicit, could you please add a separate JSON parser for that request class? This would look similar to `CommitTransactionRequestParser`

File: open-api/rest-catalog-open-api.yaml [hunk 8]
```diff
@@ -496,6 +496,58 @@ paths:
...
+
+      operationId: registerTable
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/RegisterTableRequest'
+      responses:
+        200:
+          $ref: '#/components/responses/CreateTableResponse'
```
Comment (ajantha-bhat) [inline open-api/rest-catalog-open-api.yaml:519]: should this be `RegisterTableResponse`?

File: core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java [hunk 9]
```diff
@@ -119,6 +120,11 @@ enum Route {
         LoadTableResponse.class),
     LOAD_TABLE(
         HTTPMethod.GET, "v1/namespaces/{namespace}/tables/{table}", null, LoadTableResponse.class),
+    REGISTER_TABLE(
+        HTTPMethod.POST,
+        "v1/namespaces/{namespace}/tables/register",
+        RegisterTableRequest.class,
+        LoadTableResponse.class),
```
Comment (ajantha-bhat) [inline core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:127]: do we need a new response type? Also, this is not inline with the open-api spec response type.

File: core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java [hunk 10]
```diff
@@ -761,37 +759,6 @@ public void testConversions() {
     Assert.assertEquals(ns, JdbcUtil.stringToNamespace(nsString));
   }
 
-  @Test
-  public void testRegisterTable() {
```
Comment (ajantha-bhat) [inline core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java:765]: why this is removed?

File: open-api/rest-catalog-open-api.yaml (hunk 8 above)
Comment (ajantha-bhat) [inline open-api/rest-catalog-open-api.yaml:519]: I think this should be `LoadTableResponse` as per the code implementation.

**Comment by krvikash** [comment]
(rebased with master)

**Comment by krvikash** [comment]
Thanks @nastra | @ajantha-bhat for the review. Addressed comments.

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java (hunk 1 above)
Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java]: https://github.com/apache/iceberg/pull/6591

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+
+public class RegisterTableRequestParser {
+
+  private static final String TABLE_NAME = "table-name";
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java:29]: I believe in other places we're using `identifier` instead of table name, so might be better to use the same here as well

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 12]
```diff
@@ -400,7 +402,37 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
...
+    AuthSession session = tableSession(response.config(), session(context));
+    RESTTableOperations ops =
+        new RESTTableOperations(
+            client,
+            paths.table(ident),
+            session::headers,
+            tableFileIO(context, response.config()),
+            response.tableMetadata());
+
+    return new BaseTable(ops, fullTableName(ident));
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:435]: just an FYI that we'd need to call `trackFileIO(ops);` just before the return here

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 13]
```diff
@@ -400,7 +402,37 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
...
+    Preconditions.checkArgument(
+        metadataFileLocation != null && !metadataFileLocation.isEmpty(),
+        "Invalid metadata file location: %s",
+        metadataFileLocation);
+
+    if (tableExists(context, ident)) {
+      throw new AlreadyExistsException("Table already exists: %s", ident);
+    }
+
+    RegisterTableRequest request = new RegisterTableRequest(ident.name(), metadataFileLocation);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:416]: I think it's better to track the `TableIdentifier` instead of a plain string

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  public static String toJson(RegisterTableRequest request, boolean pretty) {
+    return JsonUtil.generate(gen -> toJson(request, gen), pretty);
+  }
+
+  public static void toJson(RegisterTableRequest request, JsonGenerator gen) throws IOException {
+    Preconditions.checkArgument(null != request, "Invalid register table request: null");
+
+    gen.writeStartObject();
+
+    gen.writeStringField(TABLE_NAME, request.tableName());
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java:47]: once we track the `TableIdentifier` in the request, you can change this to 
```
gen.writeFieldName(IDENTIFIER);
TableIdentifierParser.toJson(request.identifier(), gen);
```

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+  public static RegisterTableRequest fromJson(String json) {
+    return JsonUtil.parse(json, RegisterTableRequestParser::fromJson);
+  }
+
+  public static RegisterTableRequest fromJson(JsonNode json) {
+    Preconditions.checkArgument(
+        null != json, "Cannot parse register table request from null object");
+
+    String name = JsonUtil.getString(TABLE_NAME, json);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java:61]: and this would be `TableIdentifierParser.fromJson(JsonUtil.get(IDENTIFIER, node))`

**Comment by krvikash** [comment]
Thank you @nastra for the review. Addressed comments.

**Comment by krvikash** [comment]
rebased with master and fixed CI failure.

File: core/src/main/java/org/apache/iceberg/rest/ResourcePaths.java [hunk 16]
```diff
@@ -61,6 +61,11 @@ public String tables(Namespace ns) {
     return SLASH.join("v1", prefix, "namespaces", RESTUtil.encodeNamespace(ns), "tables");
   }
 
+  public String register(Namespace ns) {
+    return SLASH.join(
+        "v1", prefix, "namespaces", RESTUtil.encodeNamespace(ns), "tables", "register");
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/ResourcePaths.java:66]: we've had some internal discussion on what the pathing should look like and `/v1/{prefix}/namespaces/{namespace}/register` makes probably the most sense here. 
Having `/v1/{prefix}/namespaces/{namespace}/tables/register` would conflict when a table is called `register`

**Comment by krvikash** [comment]
Thanks @nastra, Addressed comment

File: open-api/rest-catalog-open-api.yaml [hunk 17]
```diff
@@ -498,6 +498,58 @@ paths:
         5XX:
           $ref: '#/components/responses/ServerErrorResponse'
 
+  /v1/{prefix}/namespaces/{namespace}/tables/register:
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:501]: this seems to still have the old path

File: open-api/rest-catalog-open-api.yaml [hunk 18]
```diff
@@ -498,6 +498,58 @@ paths:
...
+      summary: Register a table in the given namespace using given metadata file location
+      description:
+        Register a table using given metadata file location.
+
+      operationId: registerTable
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/RegisterTableRequest'
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:518]: we also need to define the schema for `RegisterTableRequest`

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 19]
```diff
@@ -400,7 +403,43 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
...
+        new RESTTableOperations(
+            client,
+            paths.table(ident),
+            session::headers,
+            tableFileIO(context, response.config()),
+            response.tableMetadata());
+
+    trackFileIO(ops);
+
+    return new BaseTable(ops, fullTableName(ident));
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:442]: this also needs a parameter at the end to have proper metrics reporting: `metricsReporter(paths.metrics(ident), session::headers)`

**Review by nastra — COMMENTED** [review]
This looks almost ready, just a few small things that need to be addressed.

@danielcweeks @bryanck could you also take a look please?

**Comment by krvikash** [comment]
@nastra, Addressed comments.

**Comment by krvikash** [comment]
(fixed CI failure)

File: core/src/main/java/org/apache/iceberg/rest/RESTSerializers.java [hunk 20]
```diff
@@ -91,8 +94,12 @@ public static void registerAll(ObjectMapper mapper) {
         .addDeserializer(
             ImmutableReportMetricsRequest.class, new ReportMetricsRequestDeserializer<>())
         .addSerializer(CommitTransactionRequest.class, new CommitTransactionRequestSerializer())
+        .addDeserializer(CommitTransactionRequest.class, new CommitTransactionRequestDeserializer())
+        .addSerializer(RegisterTableRequest.class, new RegisterTableRequestSerializer<>())
+        .addDeserializer(RegisterTableRequest.class, new RegisterTableRequestDeserializer<>())
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTSerializers.java:99]: Why do we register both the interface and the implementation? Shouldn't we just need one?

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 21]
```diff
@@ -2615,6 +2616,100 @@ public void tableCreationWithoutNamespace() {
...
+    String metadataLocation = ops.current().metadataFileLocation();
+
+    catalog.dropTable(TABLE, false);
+
+    Table registeredTable = catalog.registerTable(TABLE, metadataLocation);
+    Assertions.assertThat(registeredTable).isNotNull();
+
+    TableOperations registeredTableOps = ((HasTableOperations) table).operations();
+    Assertions.assertThat(metadataLocation)
+        .isEqualTo(registeredTableOps.current().metadataFileLocation());
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2652]: This isn't necessarily true. It's fine for the catalog to import and create its own metadata file copy. I think what we want to validate is that the table is identical, not that it uses the original metadata location.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 22]
```diff
@@ -2615,6 +2616,100 @@ public void tableCreationWithoutNamespace() {
...
+
+    Map<String, String> actualProperties = Maps.newHashMap(table.properties());
+    Map<String, String> expectedProperties = Maps.newHashMap(table.properties());
+    // The nessie.commit.id for each table will be distinct, hence, it is necessary to remove it
+    // before asserting the properties.
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2657]: `CatalogTests` should not have anything specific to Nessie. Instead of doing this, I think the validation should check that the properties used to create the table are a subset of the properties in the actual table.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 23]
```diff
@@ -2615,6 +2616,100 @@ public void tableCreationWithoutNamespace() {
...
+    table.newDelete().deleteFile(FILE_A).commit();
+    table.newFastAppend().appendFile(FILE_C).commit();
+
+    TableOperations ops = ((HasTableOperations) table).operations();
+    String metadataLocation = ops.current().metadataFileLocation();
+
+    catalog.dropTable(TABLE, false);
+
+    Table registeredTable = catalog.registerTable(TABLE, metadataLocation);
+    Assertions.assertThat(registeredTable).isNotNull();
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2648]: I think this should also have a `tableExists` check/assertion just after this point. The catalog can't just return `LoadTableResponse`. It needs to have the table defined.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 24]
```diff
@@ -400,7 +403,44 @@ public void invalidateTable(SessionContext context, TableIdentifier ident) {}
...
+            .metadataFileLocation(metadataFileLocation)
+            .build();
+
+    LoadTableResponse response =
+        client.post(
+            paths.register(ident.namespace()),
+            request,
+            LoadTableResponse.class,
+            headers(context),
+            ErrorHandlers.tableErrorHandler());
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:429]: The error handler looks correct. 409 results in `AlreadyExistsException`

File: open-api/rest-catalog-open-api.yaml [hunk 25]
```diff
@@ -1913,6 +1965,17 @@ components:
           additionalProperties:
             type: string
 
+    RegisterTableRequest:
+      type: object
+      required:
+        - identifier
+        - metadata-file-location
```
Comment (rdblue) [inline open-api/rest-catalog-open-api.yaml:1972]: This should be `metadata-location`

**Comment by rdblue** [comment]
@krvikash, this is looking great! I noted a few things to fix, but overall I think the implementation is solid. Thanks for working on this!

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+    return JsonUtil.generate(gen -> toJson(request, gen), pretty);
+  }
+
+  public static void toJson(RegisterTableRequest request, JsonGenerator gen) throws IOException {
+    Preconditions.checkArgument(null != request, "Invalid register table request: null");
+
+    gen.writeStartObject();
+
+    gen.writeFieldName(IDENTIFIER);
+    TableIdentifierParser.toJson(request.identifier(), gen);
```
[the remark this answers is not in the document]
Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java:50]: As @nastra suggested (https://github.com/apache/iceberg/pull/6512#discussion_r1230706545 and  https://github.com/apache/iceberg/pull/6512#discussion_r1230708833). I modified it to send identifier.
> I think it's better to track the TableIdentifier instead of a plain string

I see that `UpdateTableRequestParser` is doing the same thing. 

https://github.com/apache/iceberg/blob/9ffb762153ce050c2d961327886ef9a41b2d6145/core/src/main/java/org/apache/iceberg/rest/requests/UpdateTableRequestParser.java#L56-L57

File: core/src/main/java/org/apache/iceberg/rest/RESTSerializers.java (hunk 20 above)
Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/rest/RESTSerializers.java:99]: Removing any one of them throws the below response. I see `ReportMetricsRequest` register both the interface and the implementation.

```
    org.apache.iceberg.exceptions.ServiceFailureException: Server error: null: {
        "servlet":"org.apache.iceberg.rest.RESTCatalogServlet-4c38cd16",
        "cause0":"com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.apache.iceberg.rest.requests.RegisterTableRequest` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information\n at [Source: (org.eclipse.jetty.server.Request$1); line: 1, column: 1]",
        "message":"com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.apache.iceberg.rest.requests.RegisterTableRequest` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information\n at [Source: (org.eclipse.jetty.server.Request$1); line: 1, column: 1]",
        "url":"/v1/namespaces/newdb/register",
        "status":"500"
        }
```

**Comment by krvikash** [comment]
Thanks @rdblue for the review, I have addressed the comments. However, there are a few open conversations.

File: core/src/main/java/org/apache/iceberg/rest/RESTSerializers.java (hunk 20 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTSerializers.java:99]: we need to register both classes unfortunately. See also https://github.com/apache/iceberg/pull/5407#discussion_r984472975 for some context and  https://github.com/apache/iceberg/pull/5911 where I attempted to fix it

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+package org.apache.iceberg.rest.requests;
+
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.rest.RESTRequest;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface RegisterTableRequest extends RESTRequest {
+
+  TableIdentifier identifier();
```
Comment (bryanck) [inline core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java:28]: The namespace in the `TableIdentifier` seems redundant with the namespace in the path, should this just be table name?

**Comment by krvikash** [comment]
(Modified `RegisterTableRequest` methods)

**Comment by krvikash** [comment]
(resolved conflicts)

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java (hunk 27 above)
Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java]: Thanks @bryanck for the review. I have updated the code to take just table-name in the request.

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (krvikash) [reply core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequestParser.java]: Hi @rdblue | @nastra, I have changed the `RegisterTableRequest` to have table-name instead identifier.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java (hunk 21 above)
Comment (krvikash) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java]: Hi @rdblue, what is your thought here?

I see in https://github.com/apache/iceberg/pull/6742, It was required to `registerTable API to avoid creating new metadata file`.

> remove the commit part in registerTable API to avoid creating new metadata file

File: core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.apache.iceberg.rest.requests;
+
+import org.apache.iceberg.rest.RESTRequest;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface RegisterTableRequest extends RESTRequest {
+
+  String tableName();
```
Comment (bryanck) [inline core/src/main/java/org/apache/iceberg/rest/requests/RegisterTableRequest.java:27]: I feel this should be named `name` to align with `CreateTableRequest`.

**Comment by krvikash** [comment]
Thanks for reviewing @bryanck, Addressed comment.

File: open-api/rest-catalog-open-api.yaml [hunk 29]
```diff
@@ -1913,6 +1965,17 @@ components:
           additionalProperties:
             type: string
 
+    RegisterTableRequest:
+      type: object
+      required:
+        - name
+        - metadata-location
+      properties:
+        identifier:
```
Comment (nastra) [inline open-api/rest-catalog-open-api.yaml:1974]:
```suggestion
        name:
```

Comment (nastra) [reply open-api/rest-catalog-open-api.yaml:1974]: this should fix the CI failure. Also make sure to go through https://github.com/apache/iceberg/blob/master/open-api/README.md#generate-python-code

File: core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java [hunk 30]
```diff
@@ -344,7 +350,13 @@ public <T extends RESTResponse> T handleRequest(
           TableIdentifier ident = identFromPathVars(vars);
           return castResponse(responseType, CatalogHandlers.loadTable(catalog, ident));
         }
-
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:347]: nit: we probably want to keep that empty line

File: core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java [hunk 31]
```diff
@@ -344,7 +350,13 @@ public <T extends RESTResponse> T handleRequest(
...
+      case REGISTER_TABLE:
+        {
+          Namespace namespace = namespaceFromPathVars(vars);
+          RegisterTableRequest request = castRequest(RegisterTableRequest.class, body);
+          return castResponse(
+              responseType, CatalogHandlers.registerTable(catalog, namespace, request));
+        }
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java:359]: nit: add an empty line after the closing } to align with all the code inside this switch block

**Review by nastra — COMMENTED** [review]
this LGTM, just a few small things to make CI happy

**Comment by krvikash** [comment]
Thanks @nastra for the review. Addressed comments.

**Review by nastra — APPROVED** [review]

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java (hunk 21 above)
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java]: I don't think that register table should be required to use the same metadata file. Some metadata may be overridden so there's no guarantee it's the same file.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 32]
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
...
+            .withSortOrder(WRITE_ORDER)
+            .withProperties(properties)
+            .create();
+
+    originalTable.newFastAppend().appendFile(FILE_A).commit();
+    originalTable.newFastAppend().appendFile(FILE_B).commit();
+    originalTable.newDelete().deleteFile(FILE_A).commit();
+    originalTable.newFastAppend().appendFile(FILE_C).commit();
+
+    TableOperations ops = ((HasTableOperations) originalTable).operations();
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2642]: Use `BaseTable` instead of `HasTableOperations`.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 33]
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
...
+    String metadataLocation = ops.current().metadataFileLocation();
+
+    catalog.dropTable(TABLE, false);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2645]: When passing a boolean, always use a comment to clarify what it means:

```java
    catalog.dropTable(TABLE, false /* do not purge */);
```

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 34]
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
...
+
+    Table registeredTable = catalog.registerTable(TABLE, metadataLocation);
+    Assertions.assertThat(registeredTable).isNotNull();
+    Assertions.assertThat(catalog.tableExists(TABLE)).as("Table must exist").isTrue();
+
+    TableOperations registeredTableOps = ((HasTableOperations) originalTable).operations();
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2651]: Use `BaseTable` here as well.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 35]
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
...
+    Assertions.assertThat(registeredTableOps.current().metadataFileLocation())
+        .isEqualTo(metadataLocation);
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2653]: I mentioned this elsewhere, but this check is unnecessary.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 36]
```diff
@@ -2615,6 +2616,95 @@ public void tableCreationWithoutNamespace() {
...
+
+    Assertions.assertThat(registeredTable.properties())
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:2655]: This needs an assertion for the table UUID.

**Review by rdblue — CHANGES_REQUESTED** [review]
Almost there! Looks like the only blocker is validating that the original metadata file is used. Just drop that assertion and this should be ready.

**Comment by krvikash** [comment]
@rdblue, thank you for reviewing. I have addressed all comments.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @krvikash! Great to have this supported in the REST catalog.

**Comment by krvikash** [comment]
Thanks @rdblue for merging. Thanks @nastra, @ajantha-bhat, @bryanck, @rdblue for the review.
